src/Microsoft.IdentityModel.Abstractions/EventDetails.cs

```diff
@@ -0,0 +1,201 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.IdentityModel.Abstractions
+{
+    /// <summary>
+    /// Details of the telemetry event.
+    /// </summary>
+    /// <remarks>
+    /// This implementation is not meant to be thread-safe. This implementation would either need to be overridden or
+    /// usage should not be concurrently operated on.
+    /// </remarks>
+    public abstract class EventDetails
+    {
+        internal Dictionary<string, DataClassification> DataClassificationMapping = new Dictionary<string, DataClassification>();
+        internal Dictionary<string, string> PropertyValues = new Dictionary<string, string>();
+        internal Dictionary<string, bool> BoolPropertyValues = new Dictionary<string, bool>();
+        internal Dictionary<string, DateTime> DateTimePropertyValues = new Dictionary<string, DateTime>();
+        internal Dictionary<string, long> LongPropertyValues = new Dictionary<string, long>();
+        internal Dictionary<string, double> DoublePropertyValues = new Dictionary<string, double>();
+        internal Dictionary<string, Guid> GuidPropertyValues = new Dictionary<string, Guid>();
+        internal Dictionary<string, DictionaryTypeEnum> PropertiesKeys = new Dictionary<string, DictionaryTypeEnum>();
+
+        /// <summary>
+        /// Name of the telemetry event, should be unique between events.
+        /// </summary>
+        public abstract string Name { get; set; }
+
+        /// <summary>
+        /// Properties which describe the event.
+        /// </summary>
+        public virtual IDictionary<string, object> Properties
+        {
+            get
+            {
+                var currentBag = new Dictionary<string, object>();
+                foreach (var kvp in PropertiesKeys)
+                {
+                    switch (kvp.Value)
+                    {
+                        case DictionaryTypeEnum.String:
+                            currentBag.Add(kvp.Key, PropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Long:
+                            currentBag.Add(kvp.Key, LongPropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Double:
+                            currentBag.Add(kvp.Key, DoublePropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Bool:
+                            currentBag.Add(kvp.Key, BoolPropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.DateTime:
+                            currentBag.Add(kvp.Key, DateTimePropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Guid:
+                            currentBag.Add(kvp.Key, GuidPropertyValues[kvp.Key]);
+                            break;
```

Comment:
Dictionary.Add() can throw exceptions when key is null or exists already.

Reply:
This is looping over PropertyKeys so at this point we already know for sure keys are unique and non-null.